### DISCUSSION TITLE
Bulk upserting with `executeRaw`

### DISCUSSION THREAD

**Discussion by jemisonf** [prisma/prisma#15452] [Q&A] [answered]
I'm having trouble figuring out the syntax for doing bulk upserts with `executeRaw`. Based on [the docs I found](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#tagged-template-helpers) my assumption was I could do something like:

```js
await prisma.$executeRaw`
  INSERT INTO foo (a, b, c)
  VALUES ${prisma.join(records.map(record => [record.a, record.b, record.c]))}
  ON DUPLICATE KEY UPDATE a = VALUES(a)`
```

But with that I get this error (this is from the real code I'm running):

> Invalid `prisma.executeRaw()` invocation: Raw query failed. Code: `1064`. Message: `You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?,?,?,?,?,?,?,?,?,?,?,?,?' at line 24`

I've tried a couple other variations without any luck, and I didn't [find anything helpful in the `sql-template-tag` docs](https://www.npmjs.com/package/sql-template-tag). Wondering if anyone has suggestions for how to get this working, or if what I'm trying to do is actually possible.

**Comment by nurul3101** [comment]
Hi @jemisonf 👋 

Does the query work as expected if you invoke it through a Database IDE?
Also, can you try adding the database schema name before the table name and see if that works?
Like instead of foo can you pass "public"."foo" and check?

**Comment by jemisonf** [comment, marked as answer]
Here's the solution I found that worked:

```typescript
await Prisma.$executeRaw`
  INSERT INTO foo (a, b, c)
  VALUES ${Prisma.join(records.map (record => Prisma.sql(`(${record.a}, ${record.b}, ${record.c})`))}
  ON DUPLICATE KEY UPDATE a = VALUES(a)`
```

i.e. wrapping one `Prisma.sql` call per record inside a `Prisma.join` call

**Reply by msgadi** [reply]
Are you using this query for mysql DB?

**Reply by glarivie** [reply]
Handle conflict (upsert) with PostgreSQL:

```sql
ON CONFLICT (a)
DO UPDATE
SET
  "a" = "excluded"."a",
  "b" = "excluded"."b",
  "c" = "excluded"."c"
;
```

**Reply by msgadi** [reply]
Thanks @glarivie
